File: frontend/src/lib/permissions.ts
// Role-based permission helpers mirroring backend/app/core/permissions.py.
export type UserRole = "admin" | "manager" | "member"

export type Permission =
  | "users:list"
  | "users:create"
  | "users:update_any"
  | "users:delete"
  | "metrics:view"
  | "profile:update_self"
  | "settings:global"

const ROLE_PERMISSIONS: Record<UserRole, readonly Permission[]> = {
  admin: [
    "users:list",
    "users:create",
    "users:update_any",
    "users:delete",
    "metrics:view",
    "profile:update_self",
    "settings:global",
  ],
  manager: ["users:list", "metrics:view", "profile:update_self"],
  member: ["profile:update_self"],
}

export function userHasPermission(
  role: UserRole | undefined,
  permission: Permission,
): boolean {
  if (!role) return false
  return ROLE_PERMISSIONS[role]?.includes(permission) ?? false
}

export function userHasAnyPermission(
  role: UserRole | undefined,
  permissions: Permission[],
): boolean {
  return permissions.some((permission) => userHasPermission(role, permission))
}

File: backend/app/core/permissions.py
# Central RBAC permission definitions and role-to-permission mapping.
import uuid
from enum import StrEnum

from app.models import User, UserRole


class Permission(StrEnum):
    USERS_LIST = "users:list"
    USERS_CREATE = "users:create"
    USERS_UPDATE_ANY = "users:update_any"
    USERS_DELETE = "users:delete"
    METRICS_VIEW = "metrics:view"
    PROFILE_UPDATE_SELF = "profile:update_self"
    SETTINGS_GLOBAL = "settings:global"
    ITEMS_LIST_ANY = "items:list_any"
    ITEMS_MANAGE_ANY = "items:manage_any"


ROLE_PERMISSIONS: dict[UserRole, frozenset[Permission]] = {
    UserRole.ADMIN: frozenset(Permission),
    UserRole.MANAGER: frozenset(
        {
            Permission.USERS_LIST,
            Permission.METRICS_VIEW,
            Permission.PROFILE_UPDATE_SELF,
        }
    ),
    UserRole.MEMBER: frozenset({Permission.PROFILE_UPDATE_SELF}),
}


def user_has_permission(user: User, permission: Permission) -> bool:
    return permission in ROLE_PERMISSIONS.get(user.role, frozenset())


def user_is_admin(user: User) -> bool:
    return user.role == UserRole.ADMIN


def user_can_manage_item(user: User, owner_id: uuid.UUID) -> bool:
    if user_has_permission(user, Permission.ITEMS_MANAGE_ANY):
        return True
    return owner_id == user.id
